Feature/mqtt client unify connection mechanism

frankdownunder opened the pull request:
This PR resolves #1375 MqttClient - Please give access to change server and port setting in runtime

slaff commented on Sming/SmingCore/Network/MqttClient.cpp:
```diff
@@ -9,6 +9,19 @@
 #include "../SmingCore.h"
 
+MqttClient::MqttClient()
+	: TcpClient((bool)false),
```
I would prefer to allow destruct to be controlled, if needed. Make the MqttClient constructor to have one optional parameter `bool autoDestruct=false` and then pass the `autoDestruct` to TcpClient(autoDestruct)

slaff commented on Sming/SmingCore/Network/MqttClient.cpp:
```diff
@@ -46,22 +63,55 @@ void MqttClient::setKeepAlive(int seconds)
 void MqttClient::setPingRepeatTime(int seconds)
 {
 	if (PingRepeatTime > keepAlive)
```
Add opening and closing curly brace. The same for the `else` block

slaff commented on Sming/SmingCore/Network/MqttClient.cpp:
```diff
@@ -46,22 +63,55 @@ void MqttClient::setKeepAlive(int seconds)
 
+bool MqttClient::connect(const URL& url, const String& clientName, uint32_t sslOptions)
+{
+	// eg						"mttqs://frank:frankspassword@frankserver:1883", "unique-client-name");
```
Remove those comments from here. Add them to the header file using the document comment syntax.

slaff commented on Sming/SmingCore/Network/MqttClient.cpp:
```diff
@@ -46,22 +63,55 @@ void MqttClient::setKeepAlive(int seconds)
+	// String Path;				empty
+	// String Query;			empty
+
+	this->server = url.Host;
```
Hm... may be it will be best to add in the class a private member called url with type URL and store the port, host, etc there. Which means that the port and server members should be removed completely from the class.

slaff commented on Sming/SmingCore/Network/MqttClient.cpp:
```diff
@@ -71,15 +121,21 @@ bool MqttClient::connect(const String& clientName, const String& username, const
+	}
 
-	if(server.length() > 0 ) {
+	if (server.length() > 0 ) 
```
For the `if` statements - please put the opening curly brace on the same line after the if condition. Separate the opening curly brace and the condition with one space. Apply that for all ifs in this class.

slaff commented on Sming/SmingCore/Network/MqttClient.cpp:
```diff
@@ -94,18 +150,31 @@ bool MqttClient::connect(const String& clientName, const String& username, const
 
 bool MqttClient::publish(String topic, String message, bool retained /* = false*/)
 {
+	if (getConnectionState() != eTCS_Ready)
```
Why is that is if condition?! Please, remove it.

slaff commented on Sming/SmingCore/Network/MqttClient.cpp:
```diff
@@ -94,18 +150,31 @@ bool MqttClient::connect(const String& clientName, const String& username, const
 
 bool MqttClient::publishWithQoS(String topic, String message, int QoS, bool retained /* = false*/, MqttMessageDeliveredCallback onDelivery /* = NULL */)
 {
+	if (getConnectionState() != eTCS_Ready)
```
Why is that is if condition?! Please, remove it.

slaff commented on Sming/SmingCore/Network/MqttClient.h:
```diff
@@ -32,16 +32,38 @@ class URL;
 	MqttClient(IPAddress serverIp, int serverPort, MqttStringSubscriptionCallback callback = NULL);
 	virtual ~MqttClient();
 
+	void setCallback(MqttStringSubscriptionCallback callback = NULL);
```
Wait a second? What is the purpose of that callback?

slaff commented on Sming/SmingCore/Network/MqttClient.h:
```diff
@@ -71,9 +93,11 @@ class MqttClient: protected TcpClient
-	int port;
+	bool privateConnect(const String& clientName, const String& username, const String& password, boolean useSsl = false, uint32_t sslOptions = 0);
+
+	String			server;
```
As already mentioned remove server, serverIp, and port and use url instead.

slaff commented on samples/MqttClient_Hello/app/application.cpp:
```diff
@@ -74,23 +74,40 @@ void onMessageReceived(String topic, String message)
+	mqtt.setCallback(onMessageReceived);
+	URL url;
+	url.Protocol = "mqtt";
+#ifdef ENABLE_SSL
```
Hm... If SSL is enabled that does not mean that you want to use SSL connection. Additionally the port is different for SSL and non-SSL connections.

frankdownunder commented:
I think I have addressed all the changes you wanted.
But there is an elefant in the room - this class still has memory leaks.  I thought at first I had introduced them so I swapped back to the develop branch.  The leaks were evident there too.  To see the leak I displayed free memory each time publishMessage() is called.  I feel it is in the underlying mqqt library, but that is just a guess.  Raise a separate issue?
I wish there was a  method I could use to track down leaks in place - overloading operator new perhaps?  Any thoughts on that?

slaff commented on samples/MqttClient_Hello/app/application.cpp:
```diff
@@ -31,17 +26,22 @@ Timer procTimer;
 // MQTT client
 // For quick check you can use: http://www.hivemq.com/demos/websocket-client/ (Connection= test.mosquitto.org:8080)
-MqttClient *mqtt;
+#ifdef USE_NEWSTYLECONNECT
```
The sample should contain only the recommended way to create an instance of the MqttClient class. Please, remove USE_NEWSTYLECONNECT and use the new recommended constructor.

slaff commented:
> But there is an elefant in the room - this class still has memory leaks. I thought at first I had introduced them so I swapped back to the develop branch. The leaks were evident there too. To see the leak I displayed free memory each time publishMessage() is called. I feel it is in the underlying mqqt library, but that is just a guess. Raise a separate issue?

Yes, please open a new issue with sample code that reproduces the problem. Once I have that information I will take care to run the code with valgrind and see if something pops up.

frankdownunder commented:
Ive taken out the code.  It was there to make sure I did not break the old
way of doing things.  We do not seem to have any unit tests ?


Cheers,

Frank Eden
P O Box 525
463 Riverview Rd
Narooma NSW 2546
Ph  02 4476 1640
      0434 197 110
clarkbaycottages.com.au
expo.naroomarotary.org.au



On 4 June 2018 at 18:54, slaff <notifications@github.com> wrote:

> *@slaff* commented on this pull request.
> ------------------------------
>
> In samples/MqttClient_Hello/app/application.cpp
> <https://github.com/SmingHub/Sming/pull/1390#discussion_r192668502>:
>
> > @@ -31,17 +26,22 @@ Timer procTimer;
>
>  // MQTT client
>  // For quick check you can use: http://www.hivemq.com/demos/websocket-client/ (Connection= test.mosquitto.org:8080)
> -MqttClient *mqtt;
> +#ifdef USE_NEWSTYLECONNECT
>
> The sample should contain only the recommended way to create an instance
> of the MqttClient class. Please, remove USE_NEWSTYLECONNECT and use the new
> recommended constructor.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/SmingHub/Sming/pull/1390#pullrequestreview-125484792>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AFK65PwbXZa2SIaZS0zyZX7OpfuwVd7kks5t5PW1gaJpZM4UPLlB>
> .
>
